parser_test.go

```diff
@@ -5348,6 +5347,185 @@ func TestInsertWithSelect(t *testing.T) {
+			query:               "select block_num(1337) from foo_1337_1 where a = block_num(10)",
+			mustFailAtResolving: true,
+		},
+		{
+			name:                "select with txn_hash()",
+			query:               "select txn_hash() from foo_1337_1",
+			mustFailAtResolving: true,
+		},
+		{
+			name:                "select with empty block_num()",
+			query:               "select block_num() from foo_1337_1",
+			mustFailAtResolving: true,
+		},
+	}
+
+	for _, it := range tests {
+		t.Run(it.name, func(tc testCase) func(t *testing.T) {
+			return func(t *testing.T) {
+				t.Parallel()
+
+				ast, err := Parse(tc.query)
+				if tc.mustFailAtParsing {
+					require.Error(t, err)
+					return
+				}
+				require.NoError(t, err)
+
+				for _, stmt := range ast.Statements {
+					resolved, err := stmt.(ReadStatement).Resolve(resolver)
+					if tc.mustFailAtResolving {
+						require.Error(t, err)
+						return
+					}
+					require.Equal(t, tc.expQuery, resolved)
+				}
+			}
+		}(it))
+	}
+}
+
+func TestCustomFunctionResolveWriteQuery(t *testing.T) {
+	t.Parallel()
+
+	type testCase struct {
+		name                string
+		query               string
+		mustFailAtParsing   bool
+		mustFailAtResolving bool
+		expQueries          []string
+	}
+
+	tests := []testCase{
+		{
+			name:       "insert with custom functions",
+			query:      "insert into foo_1337_1 values (txn_hash(), block_num())",
+			expQueries: []string{"insert into foo_1337_1 values ('0xabc', 100)"},
+		},
+		{
+			name:       "update with custom functions",
+			query:      "update foo_1337_1 SET a=txn_hash(), b=block_num() where c in (block_num(), block_num()+1)",
+			expQueries: []string{"update foo_1337_1 set a = '0xabc', b = 100 where c in (100, 100 + 1)"},
+		},
+		{
+			name:       "delete with custom functions",
+			query:      "delete from foo_1337_1 where a=block_num() and b=txn_hash()",
+			expQueries: []string{"delete from foo_1337_1 where a = 100 and b = '0xabc'"},
+		},
+		{
+			name:  "multiple queries",
+			query: "insert into foo_1337_1 values (txn_hash()); delete from foo_1337_1 where a=block_num()",
+			expQueries: []string{
+				"insert into foo_1337_1 values ('0xabc')",
+				"delete from foo_1337_1 where a = 100",
+			},
+		},
+		{
+			name:                "block_num() with integer argument",
+			query:               "delete from foo_1337_1 where a=block_num(1337)",
+			mustFailAtResolving: true,
+		},
+		{
+			name:              "block_num() with string argument",
+			query:             "delete from foo_1337_1 where a=block_num('foo')",
+			mustFailAtParsing: true,
+		},
+		{
+			name:              "txn_hash() with an integer argument",
+			query:             "insert into foo_1337_1 values (txn_hash(1))",
+			mustFailAtParsing: true,
+		},
+		{
+			name:              "txn_hash() with a string argument",
+			query:             "insert into foo_1337_1 values (txn_hash('foo'))",
+			mustFailAtParsing: true,
+		},
+	}
```

Comment:
You liked my tests? :P
We can remove them from the `go-tableland` draft PR if makes sense (or not, no strong opinions from my side).

Reply:
Yeah. The tests from `validator_test.go` is a bit redundant.

Reply:
Cool, feel free to delete  them.